File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* LCA(TreeNode* root,TreeNode* p,TreeNode* q){

        if(root == nullptr) return nullptr;
        if(root->val == p->val || root->val == q->val) return root;

        TreeNode* left = LCA(root->left,p,q);
        TreeNode* right = LCA(root->right,p,q);
        

        if(left == nullptr) return right;
        if(right == nullptr) return left;

        return root;
    }
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        return LCA(root,p,q);
    }
};

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
class Solution:
    def helper(self,root: 'TreeNode',p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if(root == None): return None
        if(root == p or root == q): return root

        left = self.helper(root.left,p,q)
        right = self.helper(root.right,p,q)

        if left and right: return root  

        if left: 
            return left
        else:
            return right
        
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        return self.helper(root, p, q) 
